app/models/voxelytics/ElasticsearchClient.scala

```diff
@@ -0,0 +1,192 @@
+        _ <- pollUntilServerStartedUp(until)
+      } yield ()
+
+    for {
+      isServerAvailableBox <- rpc(s"${conf.uri}/_cluster/health?wait_for_status=yellow&timeout=10s").request
+        .withMethod("GET")
+        .execute()
+        .flatMap(result =>
+          if (Status.isSuccessful(result.status)) {
+            Fox.successful(true)
+          } else {
+            Fox.failure(s"Unexpected error code from Elasticsearch ${result.status}.")
+        })
+        .recoverWith({
+          case e: java.net.ConnectException => {
+            logger.debug(s"Elasticsearch connection exception: $e")
+            Fox.successful(false)
+          }
+          case e =>
+            logger.error(s"Unexpected error $e")
+            Fox.failure("Unexpected error while trying to connect to Elasticsearch.", Full(e))
+        })
+      isServerAvailable <- isServerAvailableBox.toFox
+      _ <- if (!isServerAvailable) {
+        waitAndRecurse(until)
+      } else {
+        logger.info("Elasticsearch is available.")
+        Fox.successful(())
+      }
+    } yield ()
+  }
+
+  private def bootstrapIndexOnServer: Fox[Unit] =
+    for {
+      indexExists <- rpc(s"${conf.uri}/${conf.index}").head.map(_ => true).getOrElse(false)
+      _ <- if (!indexExists) {
+        logger.info("Bootstrapping Elasticsearch index for Voxelytics.")
+        rpc(s"${conf.uri}/${conf.index}").put(elasticsearchSchema)
+      } else {
+        Fox.successful(())
+      }
+    } yield ()
+
+  def queryLogs(runName: String,
+                organizationName: String,
+                taskName: Option[String],
+                minLevel: VoxelyticsLogLevel = VoxelyticsLogLevel.INFO): Fox[JsValue] = {
+
+    val levels = VoxelyticsLogLevel.values - minLevel
+    val queryStringParts = List(
+      Some(s"""vx.run_name:"$runName""""),
+      Some(s"""vx.wk_org:"$organizationName""""),
+      Some(s"level:(${levels.map(_.toString).mkString(" OR ")})"),
+      taskName.map(t => s"""vx.task_name:"$t"""")
+    ).flatten
+
+    val scrollBody = Json.obj(
+      "size" -> JsNumber(SCROLL_SIZE),
+      "query" -> Json.obj("query_string" -> Json.obj("query" -> queryStringParts.mkString(" AND "))),
+      "sort" -> Json.arr(Json.obj("@timestamp" -> Json.obj("order" -> "asc")))
+    )
+
+    val buffer = ListBuffer[JsValue]()
+
+    for {
+      _ <- serverStartupFuture
+      scroll <- rpc(s"${conf.uri}/${conf.index}/_search?scroll=1m")
+        .postJsonWithJsonResponse[JsValue, JsValue](scrollBody) ~> "Could not fetch logs"
+      scrollId = (scroll \ "_scroll_id").as[String]
+      scrollHits = (scroll \ "hits" \ "hits").as[List[JsValue]]
+      _ = buffer ++= scrollHits
+      lastScrollId <- fetchBatch(buffer, scrollId)
+      _ <- rpc(s"${conf.uri}/_search/scroll/$lastScrollId").delete()
+    } yield JsArray(buffer)
+
+  }
+
+  private def fetchBatch(buffer: ListBuffer[JsValue], scrollId: String): Fox[String] =
```

Comment:
```suggestion
  private def fetchBatchToBuffer(bufferMutable: ListBuffer[JsValue], scrollId: String): Fox[String] =
```
Since mutating passed variables is a very special case in the scala backend, I’d like to make it immediately recognizable (also goes for the other functions)